### ISSUE TITLE
compilation error - gcc does not see definitions from standard headers

### ISSUE THREAD

**Issue by cieplutki** [ebiggers/libdeflate#23]
I tried to compile libdeflate and although I thought it will be no brainer (just make && make install), it wasn't.

```
	$ make
	In file included from programs/gzip.c:28:0:
	programs/prog_util.h:145:1: error: unknown type name ‘ssize_t’

	programs/gzip.c: In function ‘restore_mode’:
	programs/gzip.c:264:2: warning: implicit declaration of function ‘fchmod’ [-Wimplicit-function-declaration]  // what are these anyway?
	programs/gzip.c: In function ‘restore_owner_and_group’:
	programs/gzip.c:273:2: warning: implicit declaration of function ‘fchown’ [-Wimplicit-function-declaration]
```

so I found it eventually that ‘ssize_t’ is declared in <unistd.h> and added it in `prog_util.h’
```
	$make
	programs/prog_util.c: In function ‘xopen_for_read’:
	programs/prog_util.c:235:2: warning: implicit declaration of function ‘posix_fadvise’ [-Wimplicit-function-declaration]
	programs/prog_util.c:235:32: error: ‘POSIX_FADV_SEQUENTIAL’ undeclared (first use in this function)
	programs/prog_util.c:235:32: note: each undeclared identifier is reported only once for each function it appears in
	programs/prog_util.c: In function ‘map_file_contents’:
	programs/prog_util.c:396:2: warning: implicit declaration of function ‘posix_madvise’ [-Wimplicit-function-declaration]
	programs/prog_util.c:396:38: error: ‘POSIX_MADV_SEQUENTIAL’ undeclared (first use in this function)

```
‘POSIX_FADV_SEQUENTIAL’ is in <fcntl.> which is included
‘POSIX_MADV_SEQUENTIAL’ is in <sys/mman.h> which also is included

So now I have no idea what to do with it - for me it should compile without problems - these values are properly defined in standard headers. Any idea how to solve it or what I do wrong?

**Comment by ebiggers** [comment]
Can you try replacing "-D_DEFAULT_SOURCE" in the Makefile with "-D_POSIX_C_SOURCE=200809"?

**Comment by cieplutki** [comment]
sure I can; before I do I solved that issue with `-std=gnu89' as sugested ny [Lorenz17](https://encode.ru/threads/2482-libdeflate-a-new-optimized-DEFLATE-implementation?s=207c75567d327e2508cbb2744399efa2&p=48134&viewfull=1#post48134)

Changing it in Makefile works, I don't get it really but it helps anyway, thanks.

**Comment by cieplutki** [comment]
Meanwhile I tried to compile it in mingw under windows and couldn't make it working, no matter what. I seems that compiler does not see definitions in standard headers although they are included. It was in `programs/gzip.c' with `_fstat64' definition which gcc didn't use eveen if I added definition from header to gzip.c. More exact details when I get in windows.

**Comment by ebiggers** [comment]
The problem was that libdeflate's Makefile, when compiling the programs, used a feature test macro that is only supported by glibc 2.19 and later.  -D_POSIX_C_SOURCE=200809 works on earlier versions too and I think it's sufficient.

I do not know what the problem with MinGW is; please open a separate issue if it's still a problem and include all the details needed to reproduce the problem, including the MinGW version and where you got it from (it works fine with my version).
